File: Proj4/Proj4A/Ackerman/Ackermann.java
public class Ackermann {

    public static int phi(int m, int n, int p) {
        if (p == 0) {
            return m + n;
        } else if (n == 0) {
            if (p == 1) {
                return 0;
            }
            else if (p == 2) {
                return 1;
            }
            else {
                return m;
            }
        } else {
            return phi(m, phi(m, n - 1, p), p - 1);
        }
    }

    public static void main(String[] args) {
        int m = 3, n = 12;

        System.out.println("phi(" + m + ", " + n + ", " + p + ") = " + phi(m, n, p));
    }
}

File: Proj4/Proj4A/Ackerman/Ackermann.pl
phi(M, N, 0, Result) :-
    Result is M + N.

phi(M, 0, 1, Result) :-
    Result is 0.

phi(M, 0, 2, Result) :-
    Result is 1.

phi(M, 0, P, Result) :-
    P > 0,
    Result is 1.

phi(M, N, P, Result) :-
    M > 0,
    N > 0,
    M1 is M - 1,
    N1 is N - 1,
    phi(M, N1, P, TempResult),
    phi(M1, TempResult, P, Result).

% (4,0) (3,7) (2,...)